File: lib/features/capture/domain/entities/issuer_rules.dart
import 'dart:convert';

import 'package:equatable/equatable.dart';

import '../../../transactions/domain/entities/transaction.dart';

/// What a rule does when it matches.
enum RuleAction {
  /// Produce a capture candidate.
  capture,

  /// Swallow the notification: it is not a movement (OTP, promo, security
  /// notice, statement, payment reminder). Nothing is persisted, nothing is
  /// counted.
  ignore,
}

/// Which part of the notification a rule's regex runs against. Declared per
/// rule because issuers are not consistent: Nu puts the amount in the title
/// and the counterparty in the body, and Nequi puts the amount mid-sentence
/// inside the body with no figure in the title at all.
enum RuleTarget { title, text, combined }

/// A single parsing rule, as declared in `assets/capture/issuer_rules.json`.
///
/// Rules are data, never code: adding an issuer format must be editing that
/// JSON plus a test case, nothing else. The Kotlin engine reads the very same
/// file — see `NotificationRuleEngine.kt`.
class IssuerRule extends Equatable {
  const IssuerRule({
    required this.ruleId,
    required this.priority,
    required this.action,
    required this.target,
    required this.pattern,
    required this.validated,
    this.entryType,
    this.captures = const <String, String>{},
    this.note,
  });

  factory IssuerRule.fromJson(Map<String, dynamic> json) {
    final String? entryTypeName = json['entryType'] as String?;
    return IssuerRule(
      ruleId: json['ruleId'] as String,
      priority: json['priority'] as int? ?? 0,
      action: (json['action'] as String? ?? 'capture') == 'ignore'
          ? RuleAction.ignore
          : RuleAction.capture,
      target: _targetFrom(json['target'] as String?),
      pattern: json['match'] as String,
      validated: json['validated'] as bool? ?? false,
      entryType: entryTypeName == null ? null : _entryTypeFrom(entryTypeName),
      captures: <String, String>{
        for (final MapEntry<String, dynamic> e
            in (json['captures'] as Map<String, dynamic>? ??
                    <String, dynamic>{})
                .entries)
          e.key: e.value as String,
      },
      note: json['note'] as String?,
    );
  }

  /// Stable id stored on every capture (`PendingCaptures.sourceRuleId`). With
  /// zero retention this is the only handle to debug a parser that starts
  /// misfiring, so it must never be reused for a different format.
  final String ruleId;

  /// Higher wins. Ignore rules sit above capture rules on purpose: an OTP that
  /// happens to contain a figure must never become a movement.
  final int priority;

  final RuleAction action;
  final RuleTarget target;

  /// The regex source. Always applied case-insensitively.
  final String pattern;

  /// `false` while nobody has seen a real notification in this format. Such a
  /// rule is a guess and is documented as such in `assets/capture/README.md`.
  final bool validated;

  /// The movement this rule produces. Null for [RuleAction.ignore] rules.
  final TransactionType? entryType;

  /// Field name -> named group in [pattern]. Supported field names: `amount`
  /// (mandatory for a capture rule), `merchant`, `last4`.
  final Map<String, String> captures;

  final String? note;

  /// Compiled form. Built lazily by the engine, not stored here, so the entity
  /// stays a value object.
  RegExp compile() => RegExp(pattern, caseSensitive: false);

  static RuleTarget _targetFrom(String? raw) => switch (raw) {
        'title' => RuleTarget.title,
        'text' => RuleTarget.text,
        _ => RuleTarget.combined,
      };

  static TransactionType _entryTypeFrom(String raw) => switch (raw) {
        'income' => TransactionType.income,
        'transfer' => TransactionType.transfer,
        // Ambiguity resolves to expense, the dominant case (HU-03).
        _ => TransactionType.expense,
      };

  @override
  List<Object?> get props => <Object?>[ruleId];
}

/// One catalogued issuer: the app whose notifications may be read, plus its
/// rules. The catalog is CLOSED (decision 2026-08-27): an app not listed here
/// can never be enabled, and therefore is never read.
class IssuerDefinition extends Equatable {
  const IssuerDefinition({
    required this.issuerId,
    required this.displayName,
    required this.packageName,
    required this.defaultCurrency,
    required this.rules,
    this.additionalPackageNames = const <String>[],
    this.packageNameValidated = false,
  });

  factory IssuerDefinition.fromJson(Map<String, dynamic> json) =>
      IssuerDefinition(
        issuerId: json['issuerId'] as String,
        displayName: json['displayName'] as String,
        packageName: json['packageName'] as String,
        defaultCurrency: json['defaultCurrency'] as String? ?? 'COP',
        additionalPackageNames: <String>[
          for (final dynamic p
              in json['additionalPackageNames'] as List<dynamic>? ??
                  <dynamic>[])
            p as String,
        ],
        packageNameValidated: json['packageNameValidated'] as bool? ?? false,
        rules: <IssuerRule>[
          for (final dynamic r in json['rules'] as List<dynamic>)
            IssuerRule.fromJson(r as Map<String, dynamic>),
        ],
      );

  final String issuerId;
  final String displayName;

  /// Primary Android package of the issuer's app.
  final String packageName;

  /// Other packages the same issuer has shipped under (rebrands, regional
  /// builds). Matching any of them counts as this issuer.
  final List<String> additionalPackageNames;

  /// `false` while the package name has not been confirmed against a real
  /// device (`adb shell pm list packages`). A wrong package is a silent
  /// failure: nothing is captured and nothing is logged, by design.
  final bool packageNameValidated;

  /// Currency assumed when the notification does not spell one out — which is
  /// always, for es-CO issuers.
  final String defaultCurrency;

  final List<IssuerRule> rules;

  /// Every package that identifies this issuer.
  List<String> get allPackageNames => <String>[
        packageName,
        ...additionalPackageNames,
      ];

  bool matchesPackage(String candidate) => allPackageNames.contains(candidate);

  @override
  List<Object?> get props => <Object?>[issuerId];
}

/// The whole rule catalog, parsed from `assets/capture/issuer_rules.json`.
class IssuerRuleSet extends Equatable {
  const IssuerRuleSet({
    required this.schemaVersion,
    required this.issuers,
    this.globalIgnoreRules = const <IssuerRule>[],
  });

  factory IssuerRuleSet.fromJson(Map<String, dynamic> json) => IssuerRuleSet(
        schemaVersion: json['schemaVersion'] as int? ?? 1,
        globalIgnoreRules: <IssuerRule>[
          for (final dynamic r
              in json['globalIgnoreRules'] as List<dynamic>? ?? <dynamic>[])
            IssuerRule.fromJson(r as Map<String, dynamic>),
        ],
        issuers: <IssuerDefinition>[
          for (final dynamic i in json['issuers'] as List<dynamic>)
            IssuerDefinition.fromJson(i as Map<String, dynamic>),
        ],
      );

  /// Parses the raw asset contents.
  factory IssuerRuleSet.fromRawJson(String raw) =>
      IssuerRuleSet.fromJson(jsonDecode(raw) as Map<String, dynamic>);

  final int schemaVersion;

  /// Applied to every issuer before its own rules: OTPs, security notices,
  /// promos, statements and payment reminders look the same everywhere.
  final List<IssuerRule> globalIgnoreRules;

  final List<IssuerDefinition> issuers;

  IssuerDefinition? issuerForPackage(String packageName) {
    for (final IssuerDefinition issuer in issuers) {
      if (issuer.matchesPackage(packageName)) {
        return issuer;
      }
    }
    return null;
  }

  IssuerDefinition? issuerById(String issuerId) {
    for (final IssuerDefinition issuer in issuers) {
      if (issuer.issuerId == issuerId) {
        return issuer;
      }
    }
    return null;
  }

  @override
  List<Object?> get props => <Object?>[schemaVersion, issuers];
}

File: android/app/src/main/kotlin/com/billetudo/app/capture/IssuerRules.kt
package com.billetudo.app.capture

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * Kotlin half of the parsing catalog.
 *
 * The rules themselves are NOT here: they live in the single source of truth
 * `assets/capture/issuer_rules.json`, which the Dart mirror engine reads too
 * (see `lib/features/capture/domain/usecases/parse_bank_notification.dart`).
 * Only the engine is duplicated. If you add a rule, add it to the JSON and to
 * the Dart test battery — never to one side alone.
 */

/** What a rule does when it matches. */
enum class RuleAction { CAPTURE, IGNORE }

/**
 * Which part of the notification the regex runs against. Declared per rule
 * because issuers disagree: Nu puts the amount in the title and the
 * counterparty in the body, Nequi hides the amount mid-sentence in the body,
 * and Google Wallet puts the merchant in the title and everything else in the
 * body.
 */
enum class RuleTarget { TITLE, TEXT, COMBINED }

/**
 * A regex whose named groups have been rewritten into plain numbered groups.
 *
 * Why: `minSdk` of this app is 24, and `Matcher.group(String)` — the accessor
 * for `(?<name>...)` — only exists from API 26. Instead of dropping named
 * groups from the JSON (which would make the rules unreadable and would break
 * parity with the Dart engine, where they work natively), the names are
 * stripped at load time and resolved to their group INDEX here. Behaviour is
 * identical on every supported Android version.
 */
class NamedRegex(pattern: String) {
    private val nameToIndex = mutableMapOf<String, Int>()
    private val regex: Regex

    init {
        val rewritten = StringBuilder()
        var groupIndex = 0
        var i = 0
        var insideCharClass = false
        while (i < pattern.length) {
            val c = pattern[i]
            when {
                // An escaped character is copied verbatim: `\(` is a literal
                // parenthesis, not a group.
                c == '\\' && i + 1 < pattern.length -> {
                    rewritten.append(c).append(pattern[i + 1])
                    i += 2
                }
                insideCharClass -> {
                    if (c == ']') insideCharClass = false
                    rewritten.append(c)
                    i++
                }
                c == '[' -> {
                    insideCharClass = true
                    rewritten.append(c)
                    i++
                }
                c == '(' -> {
                    // `(?<name>` is a capturing group; `(?<=` and `(?<!` are
                    // lookbehind and are NOT.
                    val isNamedGroup = pattern.startsWith("(?<", i) &&
                        i + 3 < pattern.length &&
                        pattern[i + 3] != '=' && pattern[i + 3] != '!'
                    when {
                        isNamedGroup -> {
                            val close = pattern.indexOf('>', i + 3)
                            require(close > 0) { "unterminated group name in: $pattern" }
                            groupIndex++
                            nameToIndex[pattern.substring(i + 3, close)] = groupIndex
                            rewritten.append('(')
                            i = close + 1
                        }
                        // Any other `(?...` construct is non-capturing.
                        pattern.startsWith("(?", i) -> {
                            rewritten.append("(?")
                            i += 2
                        }
                        else -> {
                            groupIndex++
                            rewritten.append('(')
                            i++
                        }
                    }
                }
                else -> {
                    rewritten.append(c)
                    i++
                }
            }
        }
        regex = Regex(rewritten.toString(), RegexOption.IGNORE_CASE)
    }

    fun find(input: String): MatchResult? = regex.find(input)

    /** Value of the group a rule declared under [name], or null. */
    fun group(match: MatchResult, name: String): String? {
        val index = nameToIndex[name] ?: return null
        return match.groups[index]?.value
    }
}

/** One parsing rule, as declared in the JSON. */
data class IssuerRule(
    val ruleId: String,
    val priority: Int,
    val action: RuleAction,
    val target: RuleTarget,
    /** `income` / `expense` / `transfer`; null for ignore rules. */
    val entryType: String?,
    /** field name (`amount`, `merchant`, `last4`, `currency`, `cardNetwork`) -> group name. */
    val captures: Map<String, String>,
    val regex: NamedRegex,
)

/** One catalogued issuer and its rules. */
data class IssuerDefinition(
    val issuerId: String,
    val displayName: String,
    val packageNames: List<String>,
    val defaultCurrency: String,
    val rules: List<IssuerRule>,
)

/** The whole catalog. */
data class IssuerRuleSet(
    val globalIgnoreRules: List<IssuerRule>,
    val issuers: List<IssuerDefinition>,
) {
    /**
     * The issuer that owns [packageName], or null.
     *
     * Comparison is EXACT and case-sensitive on purpose: `com.nequi.MobileApp`
     * really is capitalized and Wallet really is `walletnfcrel`, not `wallet`.
     * A typo here captures nothing and reports nothing — that silence is the
     * price of never guessing which app sent something.
     */
    fun issuerForPackage(packageName: String): IssuerDefinition? =
        issuers.firstOrNull { it.packageNames.contains(packageName) }
}

/**
 * Reads and caches the catalog from the Flutter asset bundle.
 *
 * The service may start with no Flutter engine alive, so the asset is opened
 * straight from `AssetManager` at the path Flutter packs it into.
 */
object IssuerRulesLoader {
    /** Must match the `assets:` entry in `pubspec.yaml`. */
    private const val ASSET_PATH = "flutter_assets/assets/capture/issuer_rules.json"

    @Volatile
    private var cached: IssuerRuleSet? = null

    fun load(context: Context): IssuerRuleSet {
        cached?.let { return it }
        synchronized(this) {
            cached?.let { return it }
            val raw = context.assets.open(ASSET_PATH).bufferedReader().use { it.readText() }
            val parsed = parse(JSONObject(raw))
            cached = parsed
            return parsed
        }
    }

    private fun parse(json: JSONObject): IssuerRuleSet = IssuerRuleSet(
        globalIgnoreRules = parseRules(json.optJSONArray("globalIgnoreRules")),
        issuers = buildList {
            val issuers = json.getJSONArray("issuers")
            for (i in 0 until issuers.length()) {
                val issuer = issuers.getJSONObject(i)
                add(
                    IssuerDefinition(
                        issuerId = issuer.getString("issuerId"),
                        displayName = issuer.getString("displayName"),
                        packageNames = buildList {
                            add(issuer.getString("packageName"))
                            val extra = issuer.optJSONArray("additionalPackageNames")
                            if (extra != null) {
                                for (j in 0 until extra.length()) add(extra.getString(j))
                            }
                        },
                        defaultCurrency = issuer.optString("defaultCurrency", "COP"),
                        rules = parseRules(issuer.getJSONArray("rules")),
                    ),
                )
            }
        },
    )

    private fun parseRules(array: JSONArray?): List<IssuerRule> = buildList {
        if (array == null) return@buildList
        for (i in 0 until array.length()) {
            val rule = array.getJSONObject(i)
            val captures = mutableMapOf<String, String>()
            rule.optJSONObject("captures")?.let { json ->
                json.keys().forEach { key -> captures[key] = json.getString(key) }
            }
            add(
                IssuerRule(
                    ruleId = rule.getString("ruleId"),
                    priority = rule.optInt("priority", 0),
                    action = if (rule.optString("action", "capture") == "ignore") {
                        RuleAction.IGNORE
                    } else {
                        RuleAction.CAPTURE
                    },
                    target = when (rule.optString("target", "combined")) {
                        "title" -> RuleTarget.TITLE
                        "text" -> RuleTarget.TEXT
                        else -> RuleTarget.COMBINED
                    },
                    entryType = if (rule.isNull("entryType")) null else rule.optString("entryType", null),
                    captures = captures,
                    regex = NamedRegex(rule.getString("match")),
                ),
            )
        }
    }
}
